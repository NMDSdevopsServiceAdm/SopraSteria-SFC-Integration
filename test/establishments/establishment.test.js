// this test script runs through a few various different actions on a specific establishment (registers its own establishment)

// mock the general console loggers - removes unnecessary output while running
// global.console = {
//     log: jest.fn(),
//     warn: jest.fn(),
//     error: jest.fn()
// }

const supertest = require('supertest');
const faker = require('faker');
const baseEndpoint = 'http://localhost:3000/api';
const apiEndpoint = supertest(baseEndpoint);

// mocked real postcode/location data
// http://localhost:3000/api/test/locations/random?limit=5
const locations = require('../mockdata/locations').data;
const postcodes = require('../mockdata/postcodes').data;

const registrationUtils = require('../utils/registration');

describe ("establishment", async () => {
    let cqcServices = null;
    let nonCqcServices = null;
    beforeAll(async () => {
        // clean the database
        if (process.env.CLEAN_DB) {
            await apiEndpoint.post('/test/clean')
            .send({})
            .expect(200);
        }

        // fetch the current set of CQC and non CQC services (to set main service)
        const cqcServicesResults = await apiEndpoint.get('/services/byCategory?cqc=true')
            .expect('Content-Type', /json/)
            .expect(200);
        cqcServices = cqcServicesResults.body;
            
        const nonCqcServicesResults = await apiEndpoint.get('/services/byCategory?cqc=false')
            .expect('Content-Type', /json/)
            .expect(200);
        nonCqcServices = nonCqcServicesResults.body;
    });

    beforeEach(async () => {
    });

    describe("Non CQC Establishment", async ( )=> {
        let site = null;
        let establishmentId = null;
        let loginSuccess = null;
        let authToken = null;

        beforeAll(async () => {
            site =  registrationUtils.newNonCqcSite(postcodes[1], nonCqcServices);
        });

        it("should create a non-CQC registation", async () => {
            expect(site).not.toBeNull();

            const nonCqcEstablishment = await apiEndpoint.post('/registration')
                .send([site])
                .expect('Content-Type', /json/)
                .expect(200);

            expect(nonCqcEstablishment.body.success).toEqual(1);
            expect(Number.isInteger(nonCqcEstablishment.body.establishmentId)).toEqual(true);
            establishmentId = nonCqcEstablishment.body.establishmentId;
        });

        it("should login using the given username", async () => {
            expect(establishmentId).not.toBeNull();

            // first login after registration
            const loginResponse = await apiEndpoint.post('/login')
                .send({
                    username: site.user.username,
                    password: 'password'
                })
                .expect('Content-Type', /json/)
                .expect(200);

            expect(loginResponse.body.establishment.id).toEqual(establishmentId);
            expect(loginResponse.body.establishment.isRegulated).toEqual(false);
            expect(loginResponse.body.isFirstLogin).toEqual(true);
            expect(Number.isInteger(loginResponse.body.mainService.id)).toEqual(true);

            // login a second time and confirm revised firstLogin status
            const secondLoginResponse = await apiEndpoint.post('/login')
                .send({
                    username: site.user.username,
                    password: 'password'
                })
                .expect('Content-Type', /json/)
                .expect(200);
            expect(secondLoginResponse.body.isFirstLogin).toEqual(false);
            expect(secondLoginResponse.body.establishment.name).toEqual(site.locationName);
            expect(secondLoginResponse.body.mainService.name).toEqual(site.mainService);

            loginSuccess = secondLoginResponse.body;
            
            // assert and store the auth token
            authToken = parseInt(secondLoginResponse.header.authorization);
            expect(authToken).toEqual(establishmentId);
        });

        it("should update the employer type", async () => {
            expect(authToken).not.toBeNull();
            expect(establishmentId).not.toBeNull();

            const firstResponse = await apiEndpoint.get(`/establishment/${establishmentId}/employerType`)
                .set('Authorization', authToken)
                .expect('Content-Type', /json/)
                .expect(200);
            expect(firstResponse.body.id).toEqual(establishmentId);
            expect(firstResponse.body.name).toEqual(site.locationName);
            expect(firstResponse.body.employerType).toBeNull();

            let updateResponse = await apiEndpoint.post(`/establishment/${establishmentId}/employerType`)
                .set('Authorization', authToken)
                .send({
                    employerType : "Private Sector"
                })
                .expect('Content-Type', /json/)
                .expect(200);
            expect(updateResponse.body.id).toEqual(establishmentId);
            expect(updateResponse.body.name).toEqual(site.locationName);
            expect(updateResponse.body.employerType).toEqual('Private Sector');

            const secondResponse = await apiEndpoint.get(`/establishment/${establishmentId}/employerType`)
                .set('Authorization', authToken)
                .expect('Content-Type', /json/)
                .expect(200);
            expect(secondResponse.body.id).toEqual(establishmentId);
            expect(secondResponse.body.name).toEqual(site.locationName);
            expect(secondResponse.body.employerType).toEqual('Private Sector');

            updateResponse = await apiEndpoint.post(`/establishment/${establishmentId}/employerType`)
                .set('Authorization', authToken)
                .send({
                    employerType : "Voluntary / Charity"
                })
                .expect('Content-Type', /json/)
                .expect(200);
            expect(updateResponse.body.employerType).toEqual('Voluntary / Charity');
            updateResponse = await apiEndpoint.post(`/establishment/${establishmentId}/employerType`)
                .set('Authorization', authToken)
                .send({
                    employerType : "Other"
                })
                .expect('Content-Type', /json/)
                .expect(200);
            expect(updateResponse.body.employerType).toEqual('Other');

            // now test for an unexpected employer type
            apiEndpoint.post(`/establishment/${establishmentId}/employerType`)
                .set('Authorization', authToken)
                .send({
                    employerType : "Unknown"
                })
                .expect('Content-Type', /text/)
                .expect(400)
                .end((err,res) => {
                    expect(res.text).toEqual('Unexpected employer type: Unknown');
                    expect(res.error.status).toEqual(400);
                });
            
        });
    });

    describe("CQC Establishment", async ( )=> {
        // it("should create a CQC registation", async () => {
        //     const cqcSite = registrationUtils.newCqcSite(locations[0], cqcServices);
        //     apiEndpoint.post('/registration')
        //         .send([cqcSite])
        //         .expect('Content-Type', /json/)
        //         .expect(200)
        //         .end((err, res) => {
        //             if (err) {
        //                 console.error(err);
        //             }
        //             console.log(res.body);
        //     });
        // });
    });

});