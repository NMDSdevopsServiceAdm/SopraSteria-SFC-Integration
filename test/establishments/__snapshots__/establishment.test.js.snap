// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`establishment Non CQC Establishment should update the Local Authorities Share Options 1`] = `
Array [
  Object {
    "custodianCode": 5060,
    "name": "BARKING AND DAGENHAM",
  },
  Object {
    "custodianCode": 5090,
    "name": "BARNET",
  },
  Object {
    "custodianCode": 4405,
    "name": "BARNSLEY",
  },
  Object {
    "custodianCode": 235,
    "name": "BEDFORD",
  },
  Object {
    "custodianCode": 5120,
    "name": "BEXLEY",
  },
  Object {
    "custodianCode": 4605,
    "name": "BIRMINGHAM",
  },
  Object {
    "custodianCode": 2372,
    "name": "BLACKBURN WITH DARWEN",
  },
  Object {
    "custodianCode": 2373,
    "name": "BLACKPOOL",
  },
  Object {
    "custodianCode": 4205,
    "name": "BOLTON",
  },
  Object {
    "custodianCode": 1250,
    "name": "BOURNEMOUTH",
  },
  Object {
    "custodianCode": 335,
    "name": "BRACKNELL FOREST",
  },
  Object {
    "custodianCode": 4705,
    "name": "BRADFORD MDC",
  },
  Object {
    "custodianCode": 5150,
    "name": "BRENT",
  },
  Object {
    "custodianCode": 1445,
    "name": "BRIGHTON & HOVE",
  },
  Object {
    "custodianCode": 4210,
    "name": "BURY",
  },
  Object {
    "custodianCode": 4710,
    "name": "CALDERDALE",
  },
  Object {
    "custodianCode": 5210,
    "name": "CAMDEN",
  },
  Object {
    "custodianCode": 240,
    "name": "CENTRAL BEDFORDSHIRE",
  },
  Object {
    "custodianCode": 660,
    "name": "CHESHIRE EAST",
  },
  Object {
    "custodianCode": 665,
    "name": "CHESHIRE WEST AND CHESTER",
  },
  Object {
    "custodianCode": 5030,
    "name": "CITY OF LONDON",
  },
  Object {
    "custodianCode": 5990,
    "name": "CITY OF WESTMINSTER",
  },
  Object {
    "custodianCode": 840,
    "name": "CORNWALL",
  },
  Object {
    "custodianCode": 4610,
    "name": "COVENTRY",
  },
  Object {
    "custodianCode": 5240,
    "name": "CROYDON",
  },
  Object {
    "custodianCode": 1350,
    "name": "DARLINGTON",
  },
  Object {
    "custodianCode": 1055,
    "name": "DERBY",
  },
  Object {
    "custodianCode": 1045,
    "name": "DERBYSHIRE",
  },
  Object {
    "custodianCode": 4410,
    "name": "DONCASTER",
  },
  Object {
    "custodianCode": 4615,
    "name": "DUDLEY",
  },
  Object {
    "custodianCode": 1355,
    "name": "DURHAM",
  },
  Object {
    "custodianCode": 5270,
    "name": "EALING",
  },
  Object {
    "custodianCode": 2001,
    "name": "EAST RIDING OF YORKSHIRE",
  },
  Object {
    "custodianCode": 5300,
    "name": "ENFIELD",
  },
  Object {
    "custodianCode": 4505,
    "name": "GATESHEAD",
  },
  Object {
    "custodianCode": 1620,
    "name": "GLOUCESTERSHIRE",
  },
  Object {
    "custodianCode": 5330,
    "name": "GREENWICH",
  },
  Object {
    "custodianCode": 5360,
    "name": "HACKNEY",
  },
  Object {
    "custodianCode": 650,
    "name": "HALTON",
  },
  Object {
    "custodianCode": 5390,
    "name": "HAMMERSMITH AND FULHAM",
  },
  Object {
    "custodianCode": 5450,
    "name": "HARROW",
  },
  Object {
    "custodianCode": 724,
    "name": "HARTLEPOOL",
  },
  Object {
    "custodianCode": 5480,
    "name": "HAVERING",
  },
  Object {
    "custodianCode": 1850,
    "name": "HEREFORDSHIRE",
  },
  Object {
    "custodianCode": 5510,
    "name": "HILLINGDON",
  },
  Object {
    "custodianCode": 2114,
    "name": "ISLE OF WIGHT",
  },
  Object {
    "custodianCode": 835,
    "name": "ISLES OF SCILLY",
  },
  Object {
    "custodianCode": 5570,
    "name": "ISLINGTON",
  },
  Object {
    "custodianCode": 5600,
    "name": "KENSINGTON AND CHELSEA",
  },
  Object {
    "custodianCode": 2635,
    "name": "KINGS LYNN AND WEST NORFOLK",
  },
  Object {
    "custodianCode": 2004,
    "name": "KINGSTON UPON HULL",
  },
  Object {
    "custodianCode": 5630,
    "name": "KINGSTON UPON THAMES",
  },
  Object {
    "custodianCode": 4715,
    "name": "KIRKLEES",
  },
  Object {
    "custodianCode": 4305,
    "name": "KNOWSLEY",
  },
  Object {
    "custodianCode": 5660,
    "name": "LAMBETH",
  },
  Object {
    "custodianCode": 2335,
    "name": "LANCASHIRE",
  },
  Object {
    "custodianCode": 4720,
    "name": "LEEDS",
  },
  Object {
    "custodianCode": 2465,
    "name": "LEICESTER",
  },
  Object {
    "custodianCode": 5690,
    "name": "LEWISHAM",
  },
  Object {
    "custodianCode": 2515,
    "name": "LINCOLNSHIRE",
  },
  Object {
    "custodianCode": 4310,
    "name": "LIVERPOOL",
  },
  Object {
    "custodianCode": 5180,
    "name": "LONDON BOROUGH OF BROMLEY",
  },
  Object {
    "custodianCode": 5420,
    "name": "LONDON BOROUGH OF HARINGEY",
  },
  Object {
    "custodianCode": 5540,
    "name": "LONDON BOROUGH OF HOUNSLOW",
  },
  Object {
    "custodianCode": 230,
    "name": "LUTON",
  },
  Object {
    "custodianCode": 4215,
    "name": "MANCHESTER",
  },
  Object {
    "custodianCode": 2280,
    "name": "MEDWAY",
  },
  Object {
    "custodianCode": 5720,
    "name": "MERTON",
  },
  Object {
    "custodianCode": 734,
    "name": "MIDDLESBROUGH",
  },
  Object {
    "custodianCode": 435,
    "name": "MILTON KEYNES",
  },
  Object {
    "custodianCode": 4510,
    "name": "NEWCASTLE UPON TYNE",
  },
  Object {
    "custodianCode": 5750,
    "name": "NEWHAM",
  },
  Object {
    "custodianCode": 2002,
    "name": "NORTH EAST LINCOLNSHIRE",
  },
  Object {
    "custodianCode": 2003,
    "name": "NORTH LINCOLNSHIRE",
  },
  Object {
    "custodianCode": 121,
    "name": "NORTH SOMERSET",
  },
  Object {
    "custodianCode": 4515,
    "name": "NORTH TYNESIDE",
  },
  Object {
    "custodianCode": 2935,
    "name": "NORTHUMBERLAND",
  },
  Object {
    "custodianCode": 3060,
    "name": "NOTTINGHAM",
  },
  Object {
    "custodianCode": 4220,
    "name": "OLDHAM",
  },
  Object {
    "custodianCode": 3110,
    "name": "OXFORDSHIRE",
  },
  Object {
    "custodianCode": 540,
    "name": "PETERBOROUGH",
  },
  Object {
    "custodianCode": 1160,
    "name": "PLYMOUTH",
  },
  Object {
    "custodianCode": 1255,
    "name": "POOLE",
  },
  Object {
    "custodianCode": 1775,
    "name": "PORTSMOUTH CITY COUNCIL",
  },
  Object {
    "custodianCode": 345,
    "name": "READING",
  },
  Object {
    "custodianCode": 5780,
    "name": "REDBRIDGE",
  },
  Object {
    "custodianCode": 728,
    "name": "REDCAR AND CLEVELAND",
  },
  Object {
    "custodianCode": 5810,
    "name": "RICHMOND UPON THAMES",
  },
  Object {
    "custodianCode": 4225,
    "name": "ROCHDALE",
  },
  Object {
    "custodianCode": 4415,
    "name": "ROTHERHAM",
  },
  Object {
    "custodianCode": 2470,
    "name": "RUTLAND",
  },
  Object {
    "custodianCode": 4230,
    "name": "SALFORD",
  },
  Object {
    "custodianCode": 4620,
    "name": "SANDWELL",
  },
  Object {
    "custodianCode": 4320,
    "name": "SEFTON COUNCIL",
  },
  Object {
    "custodianCode": 4420,
    "name": "SHEFFIELD",
  },
  Object {
    "custodianCode": 3245,
    "name": "SHROPSHIRE",
  },
  Object {
    "custodianCode": 350,
    "name": "SLOUGH",
  },
  Object {
    "custodianCode": 4625,
    "name": "SOLIHULL",
  },
  Object {
    "custodianCode": 1780,
    "name": "SOUTHAMPTON",
  },
  Object {
    "custodianCode": 1590,
    "name": "SOUTHEND-ON-SEA",
  },
  Object {
    "custodianCode": 119,
    "name": "SOUTH GLOUCESTERSHIRE",
  },
  Object {
    "custodianCode": 2830,
    "name": "SOUTH NORTHAMPTONSHIRE",
  },
  Object {
    "custodianCode": 3430,
    "name": "SOUTH STAFFORDSHIRE",
  },
  Object {
    "custodianCode": 4520,
    "name": "SOUTH TYNESIDE",
  },
  Object {
    "custodianCode": 5840,
    "name": "SOUTHWARK",
  },
  Object {
    "custodianCode": 4315,
    "name": "ST HELENS COUNCIL",
  },
  Object {
    "custodianCode": 4235,
    "name": "STOCKPORT",
  },
  Object {
    "custodianCode": 738,
    "name": "STOCKTON-ON-TEES",
  },
  Object {
    "custodianCode": 3530,
    "name": "SUFFOLK COASTAL",
  },
  Object {
    "custodianCode": 4525,
    "name": "SUNDERLAND",
  },
  Object {
    "custodianCode": 3640,
    "name": "SURREY HEATH",
  },
  Object {
    "custodianCode": 5870,
    "name": "SUTTON",
  },
  Object {
    "custodianCode": 4240,
    "name": "TAMESIDE",
  },
  Object {
    "custodianCode": 3240,
    "name": "TELFORD AND WREKIN",
  },
  Object {
    "custodianCode": 1165,
    "name": "TORBAY",
  },
  Object {
    "custodianCode": 5900,
    "name": "TOWER HAMLETS",
  },
  Object {
    "custodianCode": 4245,
    "name": "TRAFFORD",
  },
  Object {
    "custodianCode": 4725,
    "name": "WAKEFIELD",
  },
  Object {
    "custodianCode": 4630,
    "name": "WALSALL",
  },
  Object {
    "custodianCode": 5930,
    "name": "WALTHAM FOREST",
  },
  Object {
    "custodianCode": 5960,
    "name": "WANDSWORTH",
  },
  Object {
    "custodianCode": 655,
    "name": "WARRINGTON",
  },
  Object {
    "custodianCode": 3725,
    "name": "WARWICKSHIRE",
  },
  Object {
    "custodianCode": 340,
    "name": "WEST BERKSHIRE",
  },
  Object {
    "custodianCode": 3320,
    "name": "WEST SOMERSET",
  },
  Object {
    "custodianCode": 4250,
    "name": "WIGAN",
  },
  Object {
    "custodianCode": 3940,
    "name": "WILTSHIRE",
  },
  Object {
    "custodianCode": 355,
    "name": "WINDSOR AND MAIDENHEAD",
  },
  Object {
    "custodianCode": 4325,
    "name": "WIRRAL",
  },
  Object {
    "custodianCode": 360,
    "name": "WOKINGHAM",
  },
  Object {
    "custodianCode": 4635,
    "name": "WOLVERHAMPTON",
  },
  Object {
    "custodianCode": 425,
    "name": "WYCOMBE",
  },
  Object {
    "custodianCode": 2741,
    "name": "YORK",
  },
]
`;
